### PR TITLE
feat(github): GitHub App installation support

### PR DESCRIPTION
**Pull request by vit-zikmund**
Hello dear @rufuspollock & @athornton, long time no see :) I was missing you so much that I eventually ended up doing this beefy github auth upgrade! It adds support for GitHub App installation tokens, which is something I didn't have any knowledge about (lucky me) until I wanted to use a CICD automation (that uses a GitHub App as the governing identity) to act upon a LFS-enabled repo. Pretty nice rabbit hole I got stuck in for a several days.

Here's the feature on a rather nice silver plate. Let me know if you find any sharp corners on it, but it's proven to work in our premises. Below you'll find the the main commit description, slightly unraveling the core changes. See the updated docs for the details about this funny github token flavor.

> There is this special kind of GitHub access tokens which belong to the GitHub App installation (GitHub App installed in your GitHub organization, doing stuff on it's own behalf, like various bots). Compared to the more usual Personal Access Tokens, these tokens have entirely different way of managing their permissions and getting them from the GitHub API.
> 
> This set of changes introduces support for those tokens, which in turn led to a couple of new design decisions and some refactoring, splitting the original GithubIdentity into two subclasses, each dealing with their own authentication/authorization scheme, while still leveraging the common caching capabilities, now factored out to the parent class.
> 
> Each of the subclasses has an 'authenticate' class method, which creates an instance of the particular class, when properly authenticated against the GitHub API. This identity is then paired with its access token in a cache for later reuse.
> 
> In turn, the instance has a private '_authorize' method which provides means for obtaining the targeted org/repo permissons. These permissions are supposed to be stored in a per-identity cache via the call to '_set_permissions'. Due to the nature of the permission discovery for the new GitHub App installations, multiple repo permissions (of a single org) can be cached in a single repo. Also, when this identity has access to all organization repos, a special org/None permission is used to cover them all as a single entry.
> 
> There's also a slightly unrelated addition, which is a configurable request timeout for the GitHub API calls, starting with some rather benevolent defaults. This should bring more stability to the module, as potentially hanging requests to the API won't hog the threads forever.
> 
> Here are some more important milestones in this squash:
> * feat(github): extract username from the Basic authorization header
> 
> To be eventually used to identify an application id with an "app installation" github token.
> 
> * chore(github): move request session management under CallContext
> 
> This is needed for further design improvements where GithubIdentity objects will be able to call their specific GitHub API calls.
> 
> * chore(github): move CallContext to the module scope to hold a requests session
> 
> * feat(github): support GitHub App installation tokens
> 
> * chore(github): provide unit tests for GithubAppIdentity
> 
> * feat(github): use a timeout for GitHub API calls
> 
> * docs(github): GitHub App installation tokens
> 
> * feat(github): casually cache GitHub App repo permissions
> 
> Also: Somehow reflect on user caching preferences for the auth proxy cache, providing a hardcoded minimum for proxied entries.
> 
> GithubIdentity._set_permissions got a new flag 'casual' for writing to the main auth cache without any guarantee this will be retrieved later. This is required so casually cached permissions for GitHub Apps won't get stuck in the proxy cache.

### CODE REVIEW

**Comment by rufuspollock** [comment]
@vit-zikmund wow this is amazing. Thanks for all the hard work.

@athornton would you be up for giving this the once over and then we can merge if you think it looks good?

**Comment by athornton** [comment]
Yeah, hopefully I'll have some time tomorrow.

File: docs/source/auth-providers.md
Comment (athornton) [inline docs/source/auth-providers.md:208]: "very different ways"

File: docs/source/auth-providers.md
Comment (athornton) [inline docs/source/auth-providers.md:214]: I'd say "on" rather than "of".

File: docs/source/auth-providers.md
Comment (athornton) [inline docs/source/auth-providers.md:225]: Huh.  I think learning about `organization_administration:read` just inadvertently solved a problem I've been having in a completely separate sphere.

File: giftless/auth/github.py
Comment (athornton) [inline giftless/auth/github.py:241]: This whole bit is an interesting and elegant way to approach user caching.  I like it.

File: giftless/auth/github.py
Comment (athornton) [inline giftless/auth/github.py:353]: `outside of a with block` gets the point across, but maybe you want to do something even more pointed, like "can only be used within a context manager defining a session".

File: giftless/auth/github.py
Comment (athornton) [inline giftless/auth/github.py:389]: A little commentary here explaining what the `["0"]` is doing and why you want the first element of that (I mean, I think I can infer it from how I think the `next_url` query should look, but it's not obvious) would be gratefully received, at least by me.

File: giftless/auth/github.py
Comment (athornton) [inline giftless/auth/github.py:427]: Maybe `'ttu' function for 'TLRUCache'` because I found that hard to discern.

File: giftless/auth/github.py
Comment (athornton) [inline giftless/auth/github.py:567]: That's very clever.

File: tests/auth/test_github.py
Comment (athornton) [inline tests/auth/test_github.py:193]: Given the example, I'd suggest the id be `125678` but....

**Review by athornton — APPROVED** [review]
A few minor documentation suggestions, and one request for a comment in a place I found the action confusing, but on the whole this is fantastic.

**Comment by rufuspollock** [comment]
@vit-zikmund we'd love to merge this. wdyt of @athornton comments.

Big 🙏 to @athornton for the excellent review and 👏👏 to @vit-zikmund for the PR 🙌

**Comment by vit-zikmund** [comment]
Hi @rufuspollock I want to address them all, after all they are either sensible small improvements or nice endorsements :blush: 
I was under the water a bit last week, so couldn't get to it, but I'll promise to do it soon! Thanks for your patience :innocent:

File: docs/source/auth-providers.md
Comment (vit-zikmund) [reply docs/source/auth-providers.md:225]: I wasted an hour or so mistakenly setting the plain repo `administration: read` and banging my head why the API gives me a 403, so I think I know where you come from :wink:

File: tests/auth/test_github.py
Comment (vit-zikmund) [reply tests/auth/test_github.py:193]: Heh, I couldn't figure out why you suggest that :thinking:, care to share? But I'm doing it, as it seems it could be some funny reference or what.

**Comment by vit-zikmund** [comment]
Here are the last polishments as you requested. Thanks for the nice review!

**Comment by athornton** [comment]
@vit-zikmund I don't remember if you have merge permissions, but feel free, and if you don't, let me know and I will.

**Comment by vit-zikmund** [comment]
Nope, can't merge this myself, heh :)

**Comment by athornton** [comment]
Merged!

**Comment by vit-zikmund** [comment]
wheee! :smile:

**Comment by rufuspollock** [comment]
👏 congrats @vit-zikmund and @athornton  🙏👏

**Pull request merged**
